src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,408 @@
+            private False() { }
+        }
+
+        internal sealed class Not : AnalyzedNode
+        {
+            public readonly AnalyzedNode Operand;
+            private Not(AnalyzedNode operand) => Operand = operand;
+
+            [return: NotNullIfNotNull("operand")]
+            public static AnalyzedNode? Create(AnalyzedNode? operand)
+            {
+                Debug.Assert(!(operand is OperationEvaluation), "!(operand is OperationEvaluation)");
+                return operand switch
+                {
+                    null => null,
+                    True _ => False.Instance,
+                    False _ => True.Instance,
+                    Not p => p.Operand,
+                    Relational p => new Relational(p.Input, Negate(p.OperatorKind), p.Value),
+                    _ => new Not(operand)
+                };
+
+                static BinaryOperatorKind Negate(BinaryOperatorKind kind)
+                {
+                    return kind switch
+                    {
+                        BinaryOperatorKind.LessThan => BinaryOperatorKind.GreaterThanOrEqual,
+                        BinaryOperatorKind.GreaterThan => BinaryOperatorKind.LessThanOrEqual,
+                        BinaryOperatorKind.LessThanOrEqual => BinaryOperatorKind.GreaterThan,
+                        BinaryOperatorKind.GreaterThanOrEqual => BinaryOperatorKind.LessThan,
+                        var v => throw ExceptionUtilities.UnexpectedValue(v)
+                    };
+                }
+            }
+        }
+
+        internal abstract class Sequence : AnalyzedNode
+        {
+            public readonly ImmutableArray<AnalyzedNode> Nodes;
+            protected Sequence(ImmutableArray<AnalyzedNode> nodes) => Nodes = nodes;
+
+            public abstract AnalyzedNode Update(ArrayBuilder<AnalyzedNode> nodes);
+            public abstract AnalyzedNode Negate(ArrayBuilder<AnalyzedNode> nodes);
+
+            public static AnalyzedNode Create(bool disjunctive, ArrayBuilder<AnalyzedNode> builder)
+                => disjunctive ? OrSequence.Create(builder) : AndSequence.Create(builder);
+        }
+
+        internal sealed class AndSequence : Sequence
+        {
+            public AndSequence(ImmutableArray<AnalyzedNode> nodes) : base(nodes) { }
+
+            public override AnalyzedNode Update(ArrayBuilder<AnalyzedNode> nodes) => Create(nodes);
+            public override AnalyzedNode Negate(ArrayBuilder<AnalyzedNode> nodes) => OrSequence.Create(nodes);
+
+            public static AnalyzedNode Create(ArrayBuilder<AnalyzedNode> tests)
+            {
+                for (var i = tests.Count - 1; i >= 0; i--)
+                {
+                    switch (tests[i])
+                    {
+                        case True _:
+                            tests.RemoveAt(i);
+                            break;
+                        case False f:
+                            tests.Free();
```

Comment:
absolutely terrifying.

Reply:
i would never expect a free across method boundaries.  why can't the caller just do this.

Reply:
this portion is straight from of compiler's source. I just followed suit. See DecisitonDagBuilder.cs

Since we're further manipulating the input builder, the best way to manage the pooled object is to lend it to the factory and call Free here.

Reply:
youc an start teh method with a `using` statement which will ensure that all codepaths free the builder.  please doc that the builder is not valid after the call returns.  That approach is very unlike anything else in teh IDE, so we should call it out.